scipy/sparse/sparsetools/csr.h

```diff
@@ -556,36 +547,70 @@ void csr_toell(const I n_row,
  */
 
 /*
- * Compute the number of non-zeroes (nnz) in the result of C = A * B.
- *
+ * Compute the total number of non-zeroes (nnz) in the result of C = A * B,
+ * as well as the per-row nnz. The per-row nnz is offset by 1 so that
+ * np.cumsum (or std::inclusive_scan) will compute Cp.
  */
-template <class I>
+template <class I, class T>
 npy_intp csr_matmat_maxnnz(const I n_row,
                            const I n_col,
                            const I Ap[],
                            const I Aj[],
+                           const T Ax[],
                            const I Bp[],
-                           const I Bj[])
+                           const I Bj[],
+                           const T Bx[],
+                                 I C_row_nnz[])
 {
-    // method that uses O(n) temp storage
-    std::vector<I> mask(n_col, -1);
+    // method uses O(n * num_threads) temp storage
 
-    npy_intp nnz = 0;
-    for(I i = 0; i < n_row; i++){
-        npy_intp row_nnz = 0;
+    C_row_nnz[0] = 0;
+
+    // Group the rows into chunks, with each chunk having its own workspace.
+    // Chunks are independent and can be processed in parallel.
+    for_each_chunk(workers.par_if_flops(Ap[n_row] + Bp[n_col]),
+                   iota_iter<I>(0), iota_iter<I>(n_row), [&n_col]{
+        // create workspace for one chunk of rows
+        return make_pair(
+            std::vector<I>(n_col, -1),  // mask
+            std::vector<T>(n_col, 0));  // sums
+    }, [&](I i, auto& workspace){
+        auto& [mask, sums] = workspace;
```

Comment:
If you change the signature  to `(I first, I last)` you could iterate over the whole chunk at once, avoiding the workspace issue and allowing smarter algorithms such as computing the per thread `nnz` as you go and summing the per-thread values at the end. This would avoid you having to loop over `C_row_nnz` twice.